build_tools/check_backticks.py

```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python3 -u
+# copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
+
+"""Test script to check for invalid use of single-backticks."""
+
+__author__ = ["geetu040"]
+
+import argparse
+import ast
+import glob
+import re
+
+
+def find_py_files(folder_path):
+    """Find all Python files in a given folder path."""
+    return glob.glob(f"{folder_path}/**/*.py", recursive=True)
+
+
+def extract_docstrings(filename):
+    """Extract docstrings from a Python file."""
+    # create abstract syntax tree from the file
+    with open(filename) as f:
+        tree = ast.parse(f.read())
+
+    # walk through all nodes in the tree
+    docstrings = {}
+    for node in ast.walk(tree):
+        if (
+            isinstance(node, ast.Expr)
+            and isinstance(node.value, ast.Constant)
+            and isinstance(node.value.value, str)
+        ):
+            # if the node is an expression and
+            # its value is a constant and
+            # constant's value is a string
+            # the node represents a docstring
+            # See https://docs.python.org/3/library/ast.html#abstract-grammar
+            docstring = node.value.value
+            lineno = node.value.lineno
+            docstrings[lineno] = docstring
+
+    return docstrings
+
+
+def find_invalid_backtick_text(docstring):
+    """Find invalid backtick text in a docstring."""
+    # remove all multiple backticks to avoid interference
+    # we are looking only for invalid single-backtick
+    docstring = re.sub(r"`{2,}.*?`{2,}", "", docstring, flags=re.DOTALL)
+
+    all_backtick_text = re.findall(r"`.*?`", docstring, flags=re.DOTALL)
+    # expressions like :math:`d(x, y):= (x-y)^2` are valid cases
+    valid_backtick_text = re.findall(r":.*?:(`.*?`)", docstring, flags=re.DOTALL)
+
+    # find all the invalid backtick code snippets
+    invalid_backtick_text = set()
+    for text in all_backtick_text:
+        if text in valid_backtick_text:
+            continue
+        # rst hyperlinks are valid cases
+        if re.match(r"`.*?<http.*?>`", text, flags=re.DOTALL):
+            continue
+        invalid_backtick_text.add(text)
```

Comment:
@yarnabrina  I basically want to remove all these lines with this regex expression, would you approve of it?
```
(?<!:)(`.*?`)(?!_)
```
Hers the demo: https://regex101.com/r/5vjoZP/1

Reply:
Sorry, but I need some clarification. Why does `import sktime` need to be in double backticks? It's not a reference to an existing object, so won't link it anyway. Does it significantly alter their appearance in UI for better?

Reply:
this doesnot work with newline character in single-backtick, if flag is set to `Dot matches newline` it finds the text between 2 invalid backtick texts

Reply:
@fkiraly this is a link you may be interested in

https://github.com/numpy/numpydoc/pull/525

Reply:
too far down the rabbit hole ... 😄

Reply:
I am reading that the single-backticks are used for multiple purposes like
1. hyperlinks
2. rst roles
3. parameters
4. variable, function, class names
5. and more ...

And their use can differ and depends on the intent. Although there are some single-backtick texts that should have been in double-backtick to be rendered properly in the API reference, almost all of them fixed through this PR #6023, yet creating a script that can point them out would be pretty charlatan.

I think we should re-consider the idea of creating a build tool that tests for all invalid backticks because a script (non-intelligent) cannot elicit the invalid single-backticks based on the intent of use.

Reply:
> Sorry, but I need some clarification. Why does `import sktime` need to be in double backticks? It's not a reference to an existing object, so won't link it anyway. Does it significantly alter their appearance in UI for better?

Yes, actually UI, because in the API reference, single-back text is turned to itallic and double-backtick becomes code-highlighted

As said here: https://github.com/sktime/sktime/pull/5984#pullrequestreview-1896996308
> Btw, some docstrings use single-apostrophe (grave) by error instead of double-apostrope (double-grave) to render code snippets. If you encounter these, it would also be great to replace them.
> 
> The confusion is due to markdown havin single-grave for code snippets, but rst has double-grave.

For an example, you can see the difference in rendering of same documentation here:
- [use of single-backtick](https://www.sktime.net/en/stable/api_reference/auto_generated/sktime.alignment.dtw_numba.AlignerDtwNumba.html#sktime.alignment.dtw_numba.AlignerDtwNumba.get_test_params)
- [use of double-backtick](https://www.sktime.net/en/latest/api_reference/auto_generated/sktime.alignment.dtw_numba.AlignerDtwNumba.html#sktime.alignment.dtw_numba.AlignerDtwNumba.get_test_params)